### PR TITLE
dap python debug test at point

### PR DESCRIPTION
**Pull request by adam-r-kowalski**

### CODE REVIEW

File: dap-python.el [hunk 1]
```diff
@@ -70,6 +70,140 @@ as the pyenv version then also return nil. This works around https://github.com/
...
+
+(cl-defmethod == ((lhs lsp-symbol) (rhs lsp-symbol))
+  (and (== (lsp-symbol-name lhs) (lsp-symbol-name rhs))
+       (== (lsp-symbol-type lhs) (lsp-symbol-type rhs))
+       (== (lsp-symbol-location lhs) (lsp-symbol-location rhs))))
+
+
+
+(defun dap-python--parse-lsp-symbol (symbol)
+  (-let* (((&hash "name" "kind" "location") symbol)
```
Comment (yyoncho) [inline dap-python.el:130]: IMO we should do this either automatically or not at all. As a side note, we could use dash to add custom destructuring, for example and validate the stuff compile time with no runtime penalty. 

``` emacs-lisp
(defun dash-expand:&lsp-point (key source)
  (unless (member key '(:line :character))
    (error "Unknown key:  %s" key))
  `(gethash ,(substring (symbol-name key) 1) ,source))

;; works 
(-let [(&lsp-point :line) (ht ("line" 10)
                              ("character" 10))]
  line)

;; fails compile time
(-let [(&lsp-point :line2) (ht ("line" 10)
                               ("character" 10))]
  line2)
```

File: dap-python.el [hunk 2]
```diff
@@ -70,6 +70,140 @@ as the pyenv version then also return nil. This works around https://github.com/
...
+
+(defun dap-python--lsp-symbols-before-point (point lsp-symbols)
+  (-filter (-partial 'dap-python--lsp-symbol-before-point point) lsp-symbols))
+
+
+(defun dap-python--test-p (lsp-symbol)
+  (let ((name (lsp-symbol-name lsp-symbol)))
+    (and (== (lsp-symbol-type lsp-symbol) "Function")
+	 (>= (length name) 5)
+	 (== (substring name 0 5) "test_"))))
```
Comment (yyoncho) [inline dap-python.el:161]: `s-strarts-with?`

File: dap-python.el [hunk 3]
```diff
@@ -70,6 +70,140 @@ as the pyenv version then also return nil. This works around https://github.com/
...
+(cl-defmethod == ((lhs location) (rhs location))
+  (and (== (location-start lhs) (location-start rhs))
+       (== (location-end lhs) (location-end rhs))))
+
+
+(cl-defmethod == ((lhs lsp-symbol) (rhs lsp-symbol))
+  (and (== (lsp-symbol-name lhs) (lsp-symbol-name rhs))
+       (== (lsp-symbol-type lhs) (lsp-symbol-type rhs))
+       (== (lsp-symbol-location lhs) (lsp-symbol-location rhs))))
+
```
Comment (yyoncho) [inline dap-python.el:126]: can you keep only one empty line between methods?

**Comment by yyoncho** [comment]
Merged as it is:

`if (eq nil foo)` should be replaced with `if foo`.

Also, we should revisit the idea with structures. I accepted the PR since it is localized in dap-python.

File: dap-python.el [hunk 4]
```diff
@@ -70,6 +70,112 @@ as the pyenv version then also return nil. This works around https://github.com/
...
+(cl-defmethod dap-python--equal ((lhs string) (rhs string))
+  (string-equal lhs rhs))
+
+(cl-defmethod dap-python--equal ((lhs list) (rhs list))
+  (and (dap-python--equal (length lhs) (length rhs))
+       (-reduce (lambda (x y) (and x y)) (-zip-with 'dap-python--equal lhs rhs))))
+
+(cl-defmethod dap-python--equal ((lhs dap-python--point) (rhs dap-python--point))
+  (and (dap-python--equal (dap-python--point-line lhs) (dap-python--point-line rhs))
```
Comment (kiennq) [inline dap-python.el:103]: I think `equal` will just work here, it compares lisp structure, so all of your custom comparison should be covered with it.

**Comment by saltsucker** [comment]
Is there a special setting required for this? When I run test at point, it still does the whole buffer...?

```lisp
(defun dap-python--populate-test-at-point (conf)
  "Populate CONF with the required arguments."
  (if-let ((test (dap-python--test-at-point)))
      (plist-put conf :program (concat (buffer-file-name) test))
    (user-error "`dap-python': no test at point"))
  (plist-put conf :cwd (lsp-workspace-root))

  (dap-python--populate-start-file-args conf))

(dap-register-debug-provider "python" 'dap-python--populate-start-file-args)
(dap-register-debug-template "Python :: Run file (buffer)"
                             (list :type "python"
                                   :args ""
                                   :cwd nil
                                   :module nil
                                   :program nil
                                   :request "launch"
                                   :name "Python :: Run file (buffer)"))

(dap-register-debug-template "Python :: Run pytest (buffer)"
                             (list :type "python"
                                   :args ""
                                   :cwd nil
                                   :program nil
                                   :module "pytest"
                                   :request "launch"
                                   :name "Python :: Run pytest (buffer)"))

(dap-register-debug-provider "python-test-at-point" 'dap-python--populate-test-at-point)
(dap-register-debug-template "Python :: Run pytest (at point)"
                             (list :type "python"
                                   :args ""
                                   :program nil
                                   :module "pytest"
                                   :request "launch"
                                   :name "Python :: Run pytest (at point)"))
```
is in my `dap-python.el`.

**Pull request merged**